bluepysnap/_plotting.py

```diff
@@ -0,0 +1,363 @@
+# details.
+
+# You should have received a copy of the GNU Lesser General Public License
+# along with this library; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+"""Plotting module for the different snap objects."""
+import logging
+import numpy as np
+import pandas as pd
+
+from bluepysnap.exceptions import BluepySnapError
+from bluepysnap.sonata_constants import Node
+from bluepysnap.utils import roundrobin
+
+L = logging.getLogger(__name__)
+
+
+def _get_pyplot():
+    try:
+        import matplotlib.pyplot as plt
+    except ImportError as e:
+        msg = (
+            "Bluepysnap requirements are not installed.\n"
+            "Please pip install as follows:\n"
+            "  pip install bluepysnap[plots] --upgrade"
+        )
+        raise ImportError(str(e) + "\n\n" + msg)
+    return plt
+
+
+def spikes_firing_rate_histogram(filtered_report, time_binsize=None, ax=None):  # pragma: no cover
+    """Spike firing rate histogram.
+
+    This plot shows the number of nodes firing during a range of time.
+
+    Args:
+        time_binsize(int): bin size (milliseconds)
+        ax(matplotlib.Axis): matplotlib Axis to draw on (if not specified, pyplot.gca() is used).
+
+    Returns:
+        matplotlib.Axis: Axis containing firing rate histogram.
+
+    Notes:
+        If no axis is provided through the ax=ax keyword argument,
+        then a default layout is set using pyplot.gca().
+    """
+    # pylint: disable=too-many-locals
+    plt = _get_pyplot()
+    spike_report = filtered_report.spike_report
+
+    times = filtered_report.report.index
+    node_count = filtered_report.report[['ids', 'population']].drop_duplicates().shape[0]
+
+    if len(times) == 0:
+        raise BluepySnapError("No data to display. You should check your "
+                              "'group' query: {}.".format(spike_report.group))
+
+    time_start = np.min(times)
+    time_stop = np.max(times)
+
+    if time_binsize is None:
+        # heuristic for a nice bin size (~100 spikes per bin on average)
+        time_binsize = min(50.0, (time_stop - time_start) / ((len(times) / 100.) + 1.))
+
+    bins = np.append(np.arange(time_start, time_stop, time_binsize), time_stop)
+    hist, bin_edges = np.histogram(times, bins=bins)
+    freq = 1.0 * hist / node_count / (0.001 * time_binsize)
+
+    if ax is None:
+        ax = plt.gca()
+        ax.set_xlabel('Time [ms]')
+        ax.set_ylabel('PSTH [Hz]')
+
+    ax.plot(0.5 * (bin_edges[1:] + bin_edges[:-1]), freq, label="PSTH", drawstyle='steps-mid')
+    return ax
+
+
+def spike_raster(filtered_report, y_axis=None, ax=None):  # pragma: no cover
+    """Spike raster plot.
+
+    Shows a global overview of the circuit's firing nodes. The y axis can project either the
+    node_ids or any properties present in the different node populations.
+
+    Args:
+        y_axis (None/str): The property to display on the y axis. None is node_ids.
```

Comment:
Wouldn't it be better to make the default `y_axis="node_ids"`? Then the user doesn't have to understand what `None` means in this particular context.

Reply:
That's fair. I will use "node_id" (without "s") which is the name in specs.